### ISSUE TITLE
Replace shortcode with emoji

### ISSUE THREAD

**Issue by NoahTheDuke** [felldo/JEmoji#54]
### What feature are you proposing?

Given a string containing one or more [shortcodes](https://github.com/ikatyang/emoji-cheat-sheet?tab=readme-ov-file) (for example, `:heart:`), replace all shortcodes with their emoji equivalent (for example, ❤️).

### Do you have any additional material for your feature proposal?

Maybe this is already possible? I read through the source code and couldn't find it, but I'm not overly familiar with Java.

**Comment by felldo** [comment]
Hi, this is not as trivial as it sounds. Each platform can define their own set of shortcodes / aliases for emojis, therefore it is not directly possible to replace a shortcode with an emoji. As an example:
```json
{
  "emoji" : "🤽🏼",
  "unicode" : "\\uD83E\\uDD3D\\uD83C\\uDFFC",
  "discordAliases" : [ ":person_playing_water_polo_tone2:", ":water_polo_tone2:", ":person_playing_water_polo::skin-tone-2:", ":water_polo::skin-tone-2:" ],
  "githubAliases" : [ ],
  "slackAliases" : [ ":water_polo::skin-tone-3:" ],
  "hasFitzpatrick" : true,
  "hasHairStyle" : false,
  "version" : 3.0,
  "qualification" : "fully-qualified",
  "description" : "person playing water polo: medium-light skin tone",
  "group" : "People & Body",
  "subgroup" : "person-sport",
  "hasVariationSelectors" : false
}
```
The discord and slack alias differ in the skin tone number being used. So it is not possible to uniquely identify an emoji by alias and you would have to specify a platform. I think the best we could do in this case is to provide a list of emojis and then let the user decide which emoji they want to use. This would look somewhat like
`replaceAliases(BiFunction<String, List<Emoji>, String> replacementFunction)`
`EmojiManager.replaceAliases((alias, emojis) -> emojis.get(0).getEmoji())`

**Comment by NoahTheDuke** [comment]
Okay, no worries. Thanks.

**Comment by jacob1299** [comment]
Following up on this, would it be possible to do the opposite? Something similar like what was done in emoji-java with the EmojiParser. 

https://github.com/vdurmont/emoji-java/blob/master/src/main/java/com/vdurmont/emoji/EmojiParser.java

**Comment by felldo** [comment]
> Following up on this, would it be possible to do the opposite? Something similar like what was done in emoji-java with the EmojiParser.
> 
> https://github.com/vdurmont/emoji-java/blob/master/src/main/java/com/vdurmont/emoji/EmojiParser.java

You mean replacing emojis -> alias in a string? This is already possible. You have full control of what you want to replace an emoji with
```java
EmojiManager.replaceAllEmojis("👍 👨🏿‍🦱 😊", emoji -> emoji.getDiscordAliases().get(0))
```

**Comment by sanjay-sks** [comment]
EmojiManager.replaceAliases would be a good feature. EmojiManager.replaceAllEmojis works fine, its opposite EmojiManager.replaceAllAliases would complete the loop.  I am currently using .../com/vdurmont/emoji/EmojiParser.java which  has EmojiParser.parseToAliases and EmojiParser.parseToUnicode. That is an old library not  updated in last 5 years. I am looking forward to using this jemoji library once EmojiManager.replaceAllAliases is available.

**Comment by felldo** [comment]
@sanjay-sks @NoahTheDuke I was thinking about implementing this feature, but now I am curious: In what context do you encounter an alias in a string? Usually the alias is just an easy-to-remember UX feature that automatically converts the alias to a Unicode emoji when the associated emoji is selected from a dropdown. See for example Discord, GitHub, Slack, etc. If you fetch the text via the API, you will no longer see an alias.

**Comment by NoahTheDuke** [comment]
I no longer have a use case (I'm using a different solution) but I was looking at automatically converting markdown shortcodes to emoji unicode in blog posts for my static site generator.

**Comment by Fs00** [comment]
In [OctoDroid](https://github.com/slapperwan/gh4a) we need such a feature, because there are several situations in which the GitHub API does not replace aliases with emojis in the returned text (e.g. in repo descriptions, commit messages, label names and some more).
We're currently resorting to the same library used by @sanjay-sks, which has outdated emoji data but it's the only one that supports this use-case.

**Comment by sanjay-sks** [comment]
Sorry for late response. As Fs00 mentioned there are  a few scenarios where we need to store emoji as alias and convert it back into emoji when displaying. Wherever storage does not support 4 byte characters, this feature would be needed.

**Comment by felldo** [comment]
> Sorry for late response. As [@Fs00](https://github.com/Fs00) mentioned there are a few scenarios where we need to store emoji as alias and convert it back into emoji when displaying. Wherever storage does not support 4 byte characters, this feature would be needed.

@sanjay-sks An alias is not a safe way to store an emoji as an alias might be 2 different emojis depending on which platform you are receiving the input from. Instead the html decimal representation should be used for example. Every character should be storable in this case `a-Z0-9#&;`.

@Fs00 in this case it makes sense to have support for aliases. I was already working on it and released a new version today!
`1.7.0` https://github.com/felldo/JEmoji/releases
 Example usage can be found here https://github.com/felldo/JEmoji?tab=readme-ov-file#replacing-aliases
If you have any usage questions feel free to ask 🙂

**Comment by sanjay-sks** [comment]
Hi @felldo , 
I realize using html decimal representation would be better. I am currently using com.vdurmont.emoji.EmojiParser so my response was colored with that. Thanks for v 1.7.0 , will explore that. Will wait for your implementation on converting html decimal code into emoji.
Thanks for this up to date library.

**Comment by felldo** [comment]
> Will wait for your implementation on converting html decimal code into

@sanjay-sks This is already implemented with this new version! For that you can use
```java
EmojiManager.replaceAllEmojis("text", Emoji::getHtmlDecimalCode); // Unicode emoji -> HTML decimal notation
EmojiManager.replaceAllEmojis("text", Emoji::getUnicode, EnumSet.of(EmojiType.HTML_DECIMAL)); // HTML decimal notation -> unicode emoji
```
By default the library works with unicode emojis. Therefore, a new parameter exists to pass the types of an emoji appearence which should be handled/considered for processing

**Comment by sanjay-sks** [comment]
Is there something to replace html code on the lines of EmojiManager.replaceHtmlsCode , the reverse of EmojiManager.replaceAllEmojis("text", Emoji::getHtmlDecimalCode)?

**Comment by felldo** [comment]
> Is there something to replace html code on the lines of EmojiManager.replaceHtmlsCode , the reverse of EmojiManager.replaceAllEmojis("text", Emoji::getHtmlDecimalCode)?

Yes, check the code block in my previous comment. First line is Unicode emoji -> HTML decimal notation and the 2nd line is HTML decimal notation -> unicode emoji
